### ISSUE TITLE
EventLoopFuture.get should support cancellation

### ISSUE THREAD

**Issue by Lukasa** [apple/swift-nio#2087]
Right now `EventLoopFuture.get` is not cancellable: if users cancel a `Task` blocked in that call, nothing happens. This is a bad behaviour, and we should fix it.

**Comment by dnadoba** [comment]
Is this really a bug? I think not and we should keep the current behaviour because

### 1. It introduces invalid state transitions

This would break the general Promise/Future contract we are used to. A Promise can either be  `pending`, `succeeded` or `failed`. The only valid state transitions are from `pending` to either `succeeded` or `failed`. If it is in the `succeeded` or `failed` state, it is settled and will never change. 

If we change the behaviour as proposed, from a user perspective the promise goes from `pending` to `failed `, because `.get()` throws an error, but another call will most likely give a different result. So we would make the following transition possible: `pending` -> `failed` -> `succeeded` which is agains the promised behaviour of a Future/Promise.

### 2. it does not actually cancel the real work

The current design of Future/Promise makes that not possible. As a user, this may not be obvious and one may think that the work is actually canceled. As Futures can have multiple observers and can be chained, this can lead to some undesired results. For Example:

```swift
try {
    let allFutures = (0..<10).map {
        httpClient.get(...).flatMap {
           // write to disk
        }
    }
    try await EventLoopFuture.whenAllSucceed(allFutures, ...).get()
} catch {
    // cleanup files
}
```
- download files with `AsyncHTTPClient`
- map over the returned future to save the data asynchronously to disk
- .get() on the returned future
- cancel the surrounding task
- in the catch block, we try to cleanup the files
- files are still being written as the real work has not being canceled, we only clean up partially

---

That being said, I still very much like the general idea! However, I think we should make this obvious to the user by making this a different method with a different name. E.g. we could call it `tryToGetUntilCanceled()` or something much better.

**Comment by Lukasa** [comment]
> If we change the behaviour as proposed, from a user perspective the promise goes from `pending` to `failed` , because `.get()` throws an error, but another call will most likely give a different result. So we would make the following transition possible: `pending` -> `failed` -> `succeeded` which is agains the promised behaviour of a Future/Promise.

I think we'd be distinguishing between `cancelled` and `failed`. If you get a `CancellationError` then the lesson to draw is not that the promise failed, but that you cancelled your wait for a result. I don't think this imposes particular new costs.

Objection (2) is totally real, but it's simply not something we can solve for users. We're in a bit of a Catch-22. I think right now the behaviour of `.get` is user-hostile: users will expect that cancellation works, and it doesn't. We cannot cancel the work for them, but right now what we do is wedge the task indefinitely instead, which may not be better!

The _right_ solution in many ways is for users to clean up their work. This is not all that straightforward to do right now: you need to take a set of actions that is going to cause the promise to be completed in a timely fashion. Somewhat more annoyingly, you'll probably _also_ not throw `CancellationError` out of the failed promise.

Nonetheless, I've posed some questions in #2089 that we should consider more broadly.

**Comment by t089** [comment]
How about passing the cancellation handler to the `get()` function itself? Usually, when NIO work is cancellable you have someway to signal your NIO pipeline that you are no longer interested in the work it is doing. When somebody cancels the `Task` the correct thing to do, would be to ask the NIO pipeline to cancel/abort its work. Eg in Async HTTP client you would want to cancel the request (ie. close the connection most likely).

So let's say you have some NIO function that returns a `Work` item that holds a future for the future result of the work as well as a `cancel` function that can be called to cancel the work _best-effort_.

Here, the work is really just checking for cancellation every second, but in a HTTP client this could instead mean to close the connection, when somebody calls `cancel()`.

```swift
struct WorkCancelledError : Error {}

struct Work {
    let future: EventLoopFuture<Void>
    let cancel: () -> ()
}

func doSomeWorkOnEventLoop(_ eventLoop: EventLoop) -> Work {
    
    let cancellationFlag = NIOAtomic<Bool>.makeAtomic(value: false)
    
    let f : EventLoopFuture<Void> = eventLoop.flatSubmit {
        func tick(p: EventLoopPromise<Void>) {
            if cancellationFlag.load() {
                print("Somebody asked us to cancel, clean up work and fail our promise.")
                p.fail(WorkCancelledError())
            } else {
                print("We are busy doing our work ...")
                eventLoop.scheduleTask(in: .seconds(1)) {
                    tick(p: p)
                }
            }
        }
        
        let p = eventLoop.makePromise(of: Void.self)
        tick(p: p)
        return p.futureResult
    }
    
    return Work(future: f, cancel: { cancellationFlag.store(true) })
}

@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
func main() async {
    let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
    do {
        let el = group.next()
        
        let work = doSomeWorkOnEventLoop(el)
        
        let _ = try await work.future.getWithCancellationHandler() {
            print("The task was canceled, signal our NIO work that it is time to stop.")
            work.cancel()
        }
        
    } catch {
        print("ERROR: \(error)")
        print("Shutting down event loop group (possibly for a second time)...")
        do {
            try await group.shutdownGracefully()
        } catch {
            print("Error shutting down event loop group: \(error)")
        }
    }
}


let dg = DispatchGroup()
dg.enter()
if #available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *) {
    Task {
        
        try await withThrowingTaskGroup(of: Void.self) { group in
            group.addTask {
                await main()
            }
           
            try await Task.sleep(nanoseconds: 5_000_000_000)
            print("Cancelling the task group...")
            group.cancelAll()
        }
        
        dg.leave()
    }
} else {
    dg.leave()
}
dg.wait()
```

This would output:

```text
We are busy doing our work ...
We are busy doing our work ...
We are busy doing our work ...
We are busy doing our work ...
We are busy doing our work ...
We are busy doing our work ...
Cancelling the task group...
The task was canceled, signal our NIO work that it is time to stop.
Somebody asked us to cancel, clean up work and fail our promise.
ERROR: WorkCancelledError()
Shutting down event loop group (possibly for a second time)...
```

Note, that the error is some custom cancellation error `WorkCancelledError`, because the NIO part maybe is not yet aware of Swift concurrency types and thus uses a custom error to signal its clients that work was cancelled.

The required function on `EventLoopFuture` could look like this:

```swift
    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
    @inlinable
    public func getWithCancellationHandler(_ handler: @escaping @Sendable () -> ()) async throws -> Value {
        return try await withTaskCancellationHandler(operation: {
            return try await self.get()
        }, onCancel: {
            self.eventLoop.execute { // maybe ?
                handler()
            }
        })
    }
```

**Comment by Lukasa** [comment]
Yeah that's another perfectly valid approach. It has the upside of making it clear to the user that cancellation may not be straightforward. It has the downside of still being less attractive than `get()`, which doesn't support cancellation in any form. My worry is that `get` is an attractive nuisance.

**Comment by Lukasa** [comment]
I'm going to tag in @ktoso (who likely has concurrency-informed opinions) and @weissi (who has API-design opinions) to get some additional voices as well, because I'm really torn on the correct thing to do here.

**Comment by ktoso** [comment]
I'm on vacation this week but may sneak in some time to think about it, thanks for the ping!

**Comment by Lukasa** [comment]
We pulled together a bit of a brain trust to hash this out. Our conclusion was, for now, to do nothing.

The API proposals above are all great. Each of them solves a very real problem. Having `.get()` be able to throw cancellation errors lets users have a single cleanup path. Having explicit cancellation handlers encourages the pattern we want users to follow, where they do their own cancellation handlers. Having a separate API allows users to only pay for what they use.

Ultimately we think none of these is a silver bullet. Dealing with cancellation is hard. As a result, we think that for now we'll stick with the status quo to allow users to experiment with what works best in their own applications. This allows them (and us) to get more experience with bridging NIO to structured concurrency, which will inform whether or not we have a hole in our API surface.

**Comment by lazyjean** [comment]
> > If we change the behaviour as proposed, from a user perspective the promise goes from `pending` to `failed` , because `.get()` throws an error, but another call will most likely give a different result. So we would make the following transition possible: `pending` -> `failed` -> `succeeded` which is agains the promised behaviour of a Future/Promise.
> 
> I think we'd be distinguishing between `cancelled` and `failed`. If you get a `CancellationError` then the lesson to draw is not that the promise failed, but that you cancelled your wait for a result. I don't think this imposes particular new costs.
> 
> Objection (2) is totally real, but it's simply not something we can solve for users. We're in a bit of a Catch-22. I think right now the behaviour of `.get` is user-hostile: users will expect that cancellation works, and it doesn't. We cannot cancel the work for them, but right now what we do is wedge the task indefinitely instead, which may not be better!
> 
> The _right_ solution in many ways is for users to clean up their work. This is not all that straightforward to do right now: you need to take a set of actions that is going to cause the promise to be completed in a timely fashion. Somewhat more annoyingly, you'll probably _also_ not throw `CancellationError` out of the failed promise.
> 
> Nonetheless, I've posed some questions in #2089 that we should consider more broadly.

When I create an iOS app, cancel behavior is really useful for network module.  When user enter some ViewController, it send some request, and leave the ViewController, cancel all requests.

**Comment by Lukasa** [comment]
Absolutely, and this issue is not about whether these systems shouldn't support cancellation. The question is whether `Future.get()` _specifically_ should support cancellation, or whether users should be using `withTaskCancellationHandler` to cancel the underlying work they're doing.
